File: deeplearning4j-core/src/main/java/org/deeplearning4j/nn/conf/layers/AutoEncoder.java
/*
 *
 *  * Copyright 2015 Skymind,Inc.
 *  *
 *  *    Licensed under the Apache License, Version 2.0 (the "License");
 *  *    you may not use this file except in compliance with the License.
 *  *    You may obtain a copy of the License at
 *  *
 *  *        http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  *    Unless required by applicable law or agreed to in writing, software
 *  *    distributed under the License is distributed on an "AS IS" BASIS,
 *  *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *    See the License for the specific language governing permissions and
 *  *    limitations under the License.
 *
 */

package org.deeplearning4j.nn.conf.layers;

import lombok.*;

import org.deeplearning4j.nn.conf.Updater;
import org.deeplearning4j.nn.conf.distribution.Distribution;
import org.deeplearning4j.nn.weights.WeightInit;

/**
 *  Autoencoder.
 * Add Gaussian noise to input and learn
 * a reconstruction function.
 *
 */
@Data @NoArgsConstructor
@ToString(callSuper = true)
@EqualsAndHashCode(callSuper = true)
public class AutoEncoder extends BasePretrainNetwork {
    protected double corruptionLevel;
    protected double sparsity;

    // Builder
    private AutoEncoder(Builder builder) {
    	super(builder);
        this.corruptionLevel = builder.corruptionLevel;
        this.sparsity = builder.sparsity;
    }

    @AllArgsConstructor
    public static class Builder extends BasePretrainNetwork.Builder {
        private double corruptionLevel = 3e-1f;
        private double sparsity = 0f;

        public Builder() {}

        public Builder(double corruptionLevel) {
            this.corruptionLevel = corruptionLevel;
        }
        
        public Builder corruptionLevel(double corruptionLevel){
        	this.corruptionLevel = corruptionLevel;
        	return this;
        }
        
        public Builder sparsity(double sparsity){
        	this.sparsity = sparsity;
        	return this;
        }

        @Override
        public Builder nIn(int nIn) {
            this.nIn = nIn;
            return this;
        }

        @Override
        public Builder nOut(int nOut) {
            this.nOut = nOut;
            return this;
        }

        @Override
        public Builder activation(String activationFunction) {
            this.activationFunction = activationFunction;
            return this;
        }

        @Override
        public Builder weightInit(WeightInit weightInit) {
            this.weightInit = weightInit;
            return this;
        }

        @Override
        public Builder dropOut(double dropOut) {
            this.dropOut = dropOut;
            return this;
        }
        
        @Override
        public Builder updater(Updater updater){
        	this.updater = updater;
        	return this;
        }

        @Override
        @SuppressWarnings("unchecked")
        public AutoEncoder build() {
            return new AutoEncoder(this);
        }
        
        @Override
        public Builder dist(Distribution dist){
        	this.dist = dist;
        	return this;
        }
    }
}

File: deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark-ml/src/main/spark-1/scala/org/deeplearning4j/spark/ml/impl/AutoEncoder.scala
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.deeplearning4j.spark.ml.impl


import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.mllib.linalg.{Vector, VectorUDT, Vectors}
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.{col, udf}
import org.apache.spark.sql.types.StructType
import org.deeplearning4j.nn.conf.MultiLayerConfiguration
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork
import org.deeplearning4j.spark.ml.utils.{DatasetFacade, SchemaUtils}
import org.nd4j.linalg.factory.Nd4j


class AutoEncoder(uid: String) extends AutoEncoderWrapper[AutoEncoder, AutoEncoderModel](uid){

    def this() {
        this(Identifiable.randomUID("dl4j"))
    }

    override def mapVectorFunc = row => row.get(0).asInstanceOf[Vector]

    override def transformSchema(schema: StructType) : StructType = {
        SchemaUtils.appendColumn(schema, $(outputCol), new VectorUDT(), false)
    }

    /**
      * Fits a dataframe to the specified network configuration
      * @param dataset DataFrame
      * @return Returns an autoencoder model, which can run transformations on the vector
      */
    override def fit(dataset: DataFrame) : AutoEncoderModel = {
        val sparkdl4j = fitter(DatasetFacade.dataRows(dataset))
        new AutoEncoderModel(uid, sparkdl4j, _multiLayerConfiguration)
            .setInputCol($(inputCol))
            .setOutputCol($(outputCol))
            .setCompressedLayer($(compressedLayer))
    }

}

class AutoEncoderModel(uid: String, multiLayerNetwork: MultiLayerNetwork, multiLayerConfiguration: MultiLayerConfiguration)
    extends AutoEncoderModelWrapper[AutoEncoderModel](uid, multiLayerNetwork, multiLayerConfiguration) {

    override def udfTransformer = udf[Vector, Vector](vec => {
        val out = multiLayerNetwork.feedForwardToLayer($(compressedLayer), Nd4j.create(vec.toArray))
        val mainLayer = out.get(out.size() - 1)

        // FIXME: int cast
        val size =  mainLayer.size(1).toInt
        val values = Array.fill(size)(0.0)
        for (i <- 0 until size) {
            values(i) = mainLayer.getDouble(i.toLong)
        }
        Vectors.dense(values)
    })

    /**
      * copys an autoencoder model, including the param map
      * @param extra ParamMap
      * @return returns a copy of the autoencoder model
      */
    override def copy(extra: ParamMap) : AutoEncoderModel = {
        copyValues(new AutoEncoderModel(uid, multiLayerNetwork, multiLayerConfiguration)).setParent(parent)
    }

    /**
      * Transforms an incoming dataframe
      * @param dataFrame DataFrame
      * @return Returns a transformed dataframe.
      */
    override def transform(dataFrame: DataFrame) : DataFrame = {
        dataFrame.withColumn($(outputCol), udfTransformer(col($(inputCol))))
    }

    /**
      * Updates the schema from the new dataframe
      * @param schema StructType
      * @return Returns a struct type
      */
    override def transformSchema(schema: StructType) : StructType = {
        SchemaUtils.appendColumn(schema, $(outputCol), new VectorUDT(), false)
    }
}

object AutoEncoderModel extends AutoEncoderModelLoader